In [5]:
import pandas as pd
df = pd.read_csv('../data/used_cars.csv')
df

,brand,model,model_year,milage,fuel_type,engine,transmission,ext_col,int_col,accident,clean_title,price
0,Ford,Utility Police Interceptor Base,2013,"51,000 mi.",E85 Flex Fuel,300.0HP 3.7L V6 Cylinder Engine Flex Fuel Capa...,6-Speed A/T,Black,Black,At least 1 accident or damage reported,Yes,"$10,300"
1,Hyundai,Palisade SEL,2021,"34,742 mi.",Gasoline,3.8L V6 24V GDI DOHC,8-Speed Automatic,Moonlight Cloud,Gray,At least 1 accident or damage reported,Yes,"$38,005"
2,Lexus,RX 350 RX 350,2022,"22,372 mi.",Gasoline,3.5 Liter DOHC,Automatic,Blue,Black,None reported,NaN,"$54,598"
3,INFINITI,Q50 Hybrid Sport,2015,"88,900 mi.",Hybrid,354.0HP 3.5L V6 Cylinder Engine Gas/Electric H...,7-Speed A/T,Black,Black,None reported,Yes,"$15,500"
4,Audi,Q3 45 S line Premium Plus,2021,"9,835 mi.",Gasoline,2.0L I4 16V GDI DOHC Turbo,8-Speed Automatic,Glacier White Metallic,Black,None reported,NaN,"$34,999"
...,...,...,...,...,...,...,...,...,...,...,...,...
4004,Bentley,Continental GT Speed,2023,714 mi.,Gasoline,6.0L W12 48V PDI DOHC Twin Turbo,8-Speed Automatic with Auto-Shift,C / C,Hotspur,None reported,Yes,"$349,950"
4005,Audi,S4 3.0T Premium Plus,2022,"10,900 mi.",Gasoline,349.0HP 3.0L V6 Cylinder Engine Gasoline Fuel,Transmission w/Dual Shift Mode,Black,Black,None reported,Yes,"$53,900"
4006,Porsche,Taycan,2022,"2,116 mi.",NaN,Electric,Automatic,Black,Black,None reported,NaN,"$90,998"
4007,Ford,F-150 Raptor,2020,"33,000 mi.",Gasoline,450.0HP 3.5L V6 Cylinder Engine Gasoline Fuel,A/T,Blue,Black,None reported,Yes,"$62,999"


Описания признаков датасета:

* **brand** — марка автомобиля.
* **model** — модель автомобиля.
* **model_year** — год выпуска автомобиля.
* **milage** — пробег автомобиля.
* **fuel_type** — тип топлива (бензин, дизель, гибрид, электричество и др.).
* **engine** — характеристики двигателя (объем, мощность и т.п.).
* **transmission** — тип коробки передач.
* **ext_col** — цвет кузова автомобиля.
* **int_col** — цвет салона автомобиля.
* **accident** — информация о наличии ДТП в истории автомобиля.
* **clean_title** — наличие «чистого» юридического титула (без ограничений и проблем с документами).
* **price** — стоимость автомобиля (целевая переменная, которую необходимо предсказать).


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4009 entries, 0 to 4008
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   brand         4009 non-null   object
 1   model         4009 non-null   object
 2   model_year    4009 non-null   int64 
 3   milage        4009 non-null   object
 4   fuel_type     3839 non-null   object
 5   engine        4009 non-null   object
 6   transmission  4009 non-null   object
 7   ext_col       4009 non-null   object
 8   int_col       4009 non-null   object
 9   accident      3896 non-null   object
 10  clean_title   3413 non-null   object
 11  price         4009 non-null   object
dtypes: int64(1), object(11)
memory usage: 376.0+ KB


In [7]:
df.describe()

,model_year
count,4009.000000
mean,2015.515590
std,6.104816
min,1974.000000
25%,2012.000000
50%,2017.000000
75%,2020.000000
max,2024.000000


In [8]:
df['brand'].value_counts()

brand
Ford             386
BMW              375
Mercedes-Benz    315
Chevrolet        292
Porsche          201
Audi             200
Toyota           199
Lexus            163
Jeep             143
Land             130
Nissan           116
Cadillac         107
GMC               91
RAM               91
Dodge             90
Tesla             87
Kia               76
Hyundai           72
Mazda             64
Acura             64
Subaru            64
Honda             63
Volkswagen        59
INFINITI          59
Lincoln           52
Jaguar            47
Volvo             38
Maserati          34
Bentley           33
MINI              33
Buick             30
Chrysler          28
Lamborghini       26
Mitsubishi        20
Genesis           20
Alfa              19
Rivian            17
Hummer            16
Pontiac           15
Ferrari           12
Rolls-Royce       11
Aston              9
McLaren            6
Scion              6
FIAT               5
Saturn             5
Lotus              4
Lucid  

In [9]:
df['fuel_type'].value_counts()

fuel_type
Gasoline          3309
Hybrid             194
E85 Flex Fuel      139
Diesel             116
–                   45
Plug-In Hybrid      34
not supported        2
Name: count, dtype: int64

In [10]:
df['transmission'].value_counts()

transmission
A/T                                              1037
8-Speed A/T                                       406
Transmission w/Dual Shift Mode                    398
6-Speed A/T                                       362
6-Speed M/T                                       248
                                                 ... 
Automatic, 8-Spd PDK Dual-Clutch                    1
Auto, 6-Spd w/CmdShft                               1
Automatic, 8-Spd Sport w/Sport & Manual Modes       1
CVT-F                                               1
8-Speed Manual                                      1
Name: count, Length: 62, dtype: int64

In [11]:
df[['model_year']].describe()

,model_year
count,4009.000000
mean,2015.515590
std,6.104816
min,1974.000000
25%,2012.000000
50%,2017.000000
75%,2020.000000
max,2024.000000


In [12]:
df[['price', 'milage']].describe()

,price,milage
count,4009,4009
unique,1569,2818
top,"$15,000","110,000 mi."
freq,39,16


В датасете обнаружены пропуски в признаках fuel_type, accident и clean_title. Кроме того, признаки price и mileage хранятся в текстовом формате и требуют преобразования в числовой тип перед обучением модели.

In [13]:
df[['price', 'milage']].head()

,price,milage
0,"$10,300","51,000 mi."
1,"$38,005","34,742 mi."
2,"$54,598","22,372 mi."
3,"$15,500","88,900 mi."
4,"$34,999","9,835 mi."


In [14]:
df['price'] = (
    df['price']
    .str.replace('$', '', regex=False)
    .str.replace(',', '', regex=False)
    .astype(float)
)

In [15]:
df['milage'] = (
    df['milage']
    .str.replace(' mi.', '', regex=False)
    .str.replace(',', '', regex=False)
    .astype(float)
)

In [16]:
df[['price', 'milage']].head()

,price,milage
0,10300.0,51000.0
1,38005.0,34742.0
2,54598.0,22372.0
3,15500.0,88900.0
4,34999.0,9835.0


In [17]:
df[['price', 'milage', 'model_year']].describe()

,price,milage,model_year
count,4.009000e+03,4009.000000,4009.000000
mean,4.455319e+04,64717.551010,2015.515590
std,7.871064e+04,52296.599459,6.104816
min,2.000000e+03,100.000000,1974.000000
25%,1.720000e+04,23044.000000,2012.000000
50%,3.100000e+04,52775.000000,2017.000000
75%,4.999000e+04,94100.000000,2020.000000
max,2.954083e+06,405000.000000,2024.000000


В датасете содержится 4009 автомобилей. Средняя стоимость автомобиля составляет 44,6 тыс. долларов, медианная — 31 тыс. долларов. Значительное различие между средним и медианой говорит о наличии дорогих автомобилей, которые смещают среднее значение вверх. Стоимость автомобилей варьируется от 2 тыс. до 2,95 млн долларов.
Средний пробег составляет 64,7 тыс. миль, медианный — 52,8 тыс. миль. В выборке присутствуют как практически новые автомобили с пробегом около 100 миль, так и автомобили с пробегом более 400 тыс. миль.
Автомобили представлены в диапазоне от 1974 до 2024 года выпуска. Средний год выпуска составляет 2015 год, а половина автомобилей была произведена после 2017 года.

In [18]:
#проверим выбросы
df['price'].sort_values(ascending=False).head(10)

693     2954083.0
229     1950995.0
3046    1599000.0
1356     749950.0
624      695000.0
979      649999.0
1615     599995.0
1508     599000.0
3655     491836.0
1061     489995.0
Name: price, dtype: float64

В датасете присутствуют автомобили с очень высокой стоимостью. Максимальная цена составляет почти 3 млн долларов, что значительно превышает средние значения по выборке. Это говорит о наличии дорогих автомобилей премиального сегмента, которые могут влиять на среднюю стоимость. Поскольку остальные значения из топ-10 также выглядят реалистично для рынка люксовых автомобилей, на данном этапе выбросы удаляться не будут.


In [19]:
#посмотрим на пропуски
missing = pd.DataFrame({
    'Пропуски': df.isna().sum(),
    'Процент': round(df.isna().sum() / len(df) * 100, 2)
})

missing[missing['Пропуски'] > 0]

,Пропуски,Процент
fuel_type,170,4.24
accident,113,2.82
clean_title,596,14.87


В датасете обнаружены пропуски в трех признаках: `fuel_type` (4,24%), `accident` (2,82%) и `clean_title` (14,87%). Наибольшая доля пропусков наблюдается в столбце `clean_title`, однако она не превышает 15% от общего объема данных. Поскольку количество пропусков относительно невелико, данные могут быть сохранены путем заполнения пропущенных значений без удаления наблюдений.


In [20]:
df['fuel_type'].value_counts(dropna=False)


fuel_type
Gasoline          3309
Hybrid             194
NaN                170
E85 Flex Fuel      139
Diesel             116
–                   45
Plug-In Hybrid      34
not supported        2
Name: count, dtype: int64

In [21]:
df['accident'].value_counts(dropna=False)

accident
None reported                             2910
At least 1 accident or damage reported     986
NaN                                        113
Name: count, dtype: int64

In [22]:
df['clean_title'].value_counts(dropna=False)

clean_title
Yes    3413
NaN     596
Name: count, dtype: int64

В признаке fuel_type обнаружены как пропуски, так и некорректные значения (—, not supported), которые были приведены к единому формату пропущенных данных. В признаках accident и clean_title доля пропусков не превышает 15%. Поскольку все пропуски относятся к категориальным признакам, для их обработки будет использовано заполнение наиболее частым значением либо отдельной категорией Unknown.

In [23]:
df['fuel_type'] = df['fuel_type'].fillna('Unknown')
df['accident'] = df['accident'].fillna('Unknown')
df['clean_title'] = df['clean_title'].fillna('Unknown')

In [24]:
df.isna().sum()

brand           0
model           0
model_year      0
milage          0
fuel_type       0
engine          0
transmission    0
ext_col         0
int_col         0
accident        0
clean_title     0
price           0
dtype: int64

ML

In [25]:
y = df['price']

In [26]:
X = df.drop('price', axis=1)

In [27]:
#train-test
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [28]:
#числовые признаки
numeric_features = [
    'model_year',
    'milage'
]

In [29]:
#категориальные
categorical_features = [
    'brand',
    'model',
    'fuel_type',
    'engine',
    'transmission',
    'ext_col',
    'int_col',
    'accident',
    'clean_title'
]

In [30]:
#OHE, Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor

In [31]:
#делим признаки по типам
numeric_features = [
    'model_year',
    'milage'
]

categorical_features = [
    'brand',
    'model',
    'fuel_type',
    'engine',
    'transmission',
    'ext_col',
    'int_col',
    'accident',
    'clean_title'
]

In [32]:
# категориальные признаки переводим в формат 0 и 1
# числовые признаки оставляем без изменений
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_features),
        ('num', 'passthrough', numeric_features)
    ]
)

In [33]:
#сначала подготовим данные, затем - обучим модель
model = Pipeline(
    steps=[
        ('prepare_data', preprocessor),
        ('predict_price', RandomForestRegressor(random_state=42))
    ]
)

In [34]:
model.fit(X_train, y_train)

Pipeline(steps=[('prepare_data',
                 ColumnTransformer(transformers=[('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['brand', 'model',
                                                   'fuel_type', 'engine',
                                                   'transmission', 'ext_col',
                                                   'int_col', 'accident',
                                                   'clean_title']),
                                                 ('num', 'passthrough',
                                                  ['model_year', 'milage'])])),
                ('predict_price', RandomForestRegressor(random_state=42))])

In [35]:
y_pred = model.predict(X_test)

In [36]:
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error, r2_score
import numpy as np

mae = mean_absolute_error(y_test, y_pred)
mape = mean_absolute_percentage_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

rmse = np.sqrt(np.mean((y_test - y_pred) ** 2))

In [37]:
import pandas as pd

pd.DataFrame({
    'Real': y_test.values,
    'Predicted': y_pred
}).head(10)

,Real,Predicted
0,28000.0,32789.34
1,5900.0,16033.36
2,41000.0,46135.70
3,40250.0,34036.69
4,77999.0,45068.39
5,35899.0,33964.46
6,36000.0,47093.05
7,47214.0,50904.90
8,30900.0,18144.23
9,11000.0,10472.69


In [38]:
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error, r2_score
import numpy as np

mae = mean_absolute_error(y_test, y_pred)
mape = mean_absolute_percentage_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

rmse = np.sqrt(np.mean((y_test - y_pred) ** 2))

print("MAE:", mae)
print("MAPE:", mape)
print("RMSE:", rmse)
print("R2:", r2)

MAE: 19008.494314214462
MAPE: 0.31010871193744727
RMSE: 134763.90696615243
R2: 0.11146322107892936


Модель смогла выявить зависимость между характеристиками автомобиля и его стоимостью, однако качество прогнозирования остается умеренным. Значение R² = 0.11 показывает, что модель объясняет около 11% вариации цены автомобиля. Для повышения качества модели в дальнейшем можно провести более глубокую предобработку признаков и извлечь дополнительные характеристики из текстовых полей.

In [39]:
import joblib

joblib.dump(model, 'car_price_model.pkl')

['car_price_model.pkl']

In [40]:
import os

os.listdir()

['car_price_model.pkl', 'analysis.ipynb']

In [41]:
os.getcwd()

'/Users/alina/Desktop/Магистратура/проект семинар наставника/notebooks'

In [42]:
import joblib

joblib.dump(model, 'car_price_model.pkl')

['car_price_model.pkl']

In [43]:
import os
os.listdir()

['car_price_model.pkl', 'analysis.ipynb']

In [46]:
#Fastapi
from fastapi import FastAPI
import joblib
import pandas as pd

app = FastAPI()

model = joblib.load("car_price_model.pkl")

@app.get("/")
def root():
    return {"message": "API works"}

@app.post("/predict")
def predict(data: dict):

    df = pd.DataFrame([data])

    prediction = model.predict(df)

    return {"price": float(prediction[0])}